### PR TITLE
Draft: Add filter rule( gcc 9 and older does not support C++ 20 )

### PR DESCRIPTION
**Pull request by YuriiPerets**

### CODE REVIEW

File: src/bashi/versions.py [hunk 1]
```diff
@@ -123,6 +140,16 @@ def __str__(self) -> str:
 ]
 NVCC_GCC_MAX_VERSION.sort(reverse=True)
 
+
+# define the maximum supported gcc version for a specific cxx version
+# the latest supported cxx version must be added, even if the supported gcc version does not
+# increase
+GCC_CXX_SUPPORT: List[GccCxxSupport] = [
+    GccCxxSupport("14", "23"),
```
Comment (SimeonEhrig) [inline src/bashi/versions.py:148]: I thought about this again. Actual the GCC 15 (not released yet) also adds missing C++23 and maybe the GCC 16 will too. So my idea is not good. Therefore we remove the entry for C++23 and replace it with C++17.
```suggestion
    GccCxxSupport("8", "17"),
```

The filter rule should still work. You need only to modify the tests. Sorry for the extra work.

File: src/bashi/versions.py [hunk 2]
```diff
@@ -123,6 +140,16 @@ def __str__(self) -> str:
...
+    #   GccCxxSupport("18", "26"),
```
Comment (SimeonEhrig) [inline src/bashi/versions.py:149]: Remove the comment.

File: src/bashi/filter_software_dependency.py [hunk 3]
```diff
@@ -156,4 +156,58 @@ def software_dependency_filter(
                         f"{__ubuntu_version_to_string(row[UBUNTU].version)}",
                     )
                     return False
+
+        # Rule: d5
+        # remove all unsupported gcc cxx version combinations
+        """
+    if DEVICE_COMPILER in row and row[DEVICE_COMPILER].name == GCC:
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:163]: Don't forget to remove this.

File: src/bashi/filter_software_dependency.py [hunk 4]
```diff
@@ -156,4 +156,58 @@ def software_dependency_filter(
...
+                                        output,
+                                        "host compiler"
+                                        f" gcc {row[HOST_COMPILER].version} "
+                                        f"does not support cxx {row[CXX_STANDARD].version}"
+                                    )
+                                    return False
+                                break    
+          """
+
+    for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:196]: Just as note, could be that you need to modify the rule, if you implement the C++ support filter rule for `nvcc`. Because with the `nvcc` filter rule, the C++ support of the host compiler depends on the device compiler. Small example:
If the host and device compiler is GCC 11, it supports C++20. If the host compiler is GCC 11 and the device Compiler is NVCC 11.8, only up to C++17 is allowed.
On the other side, the rule already removes invalid combinations which can never happen. If we check `GCC 11 + C++ 23`, it will never pass independent if the device compiler is GCC or NVCC. Therefore we keep it.

I only write down my thoughts in case you get a `covertable.exceptions.InvalidCondition` exception when you implement the nvcc C++ standard filter rule, that you have an point for under investigating.

File: tests/test_results.py [hunk 5]
```diff
@@ -2644,3 +2552,45 @@ def test_remove_unsupported_cuda_versions_for_ubuntu(self):
             expected_results,
             create_diff_parameter_value_pairs(test_param_value_pairs, expected_results),
         )
+
+    def test_remove_unsupported_cxx_versions_for_gcc(self):
+        test_param_value_pairs: List[ParameterValuePair] = parse_expected_val_pairs2(
+            [
```
Comment (SimeonEhrig) [inline tests/test_results.py:2558]: The corner cases `((HOST_COMPILER, GCC, 9), (CXX_STANDARD, 19)),` and `((HOST_COMPILER, GCC, 9), (CXX_STANDARD, 20)),` and `((HOST_COMPILER, GCC, 8), (CXX_STANDARD, 16)),` and `((HOST_COMPILER, GCC, 8), (CXX_STANDARD, 17)),` are missing.

**Review by SimeonEhrig — APPROVED** [review]

File: src/bashi/results.py [hunk 6]
```diff
@@ -908,3 +909,35 @@ def _remove_unsupported_cuda_versions_for_ubuntu(
...
             value_max_version2_inclusive=False,
         )
+
+
+def _remove_unsupported_cxx_versions_for_gcc(
+    parameter_value_pairs: List[ParameterValuePair],
+    removed_parameter_value_pairs: List[ParameterValuePair],
+):
+    for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
+        remove_parameter_value_pairs_ranges(
```
Comment (SimeonEhrig) [inline src/bashi/results.py:919]: I think the usage of `remove_parameter_value_pairs_ranges()` is not correct yet. Because in the end, we should be able to iterate of the list `GCC_CXX_SUPPORT` and call the function `remove_parameter_value_pairs_ranges()` with each combination of the `GCC` version and `CXX_STANDARD`.

Comment (YuriiPerets) [reply src/bashi/results.py:978]: Yes, it's a temporary solution, I wanted to write the right test_results first, and then change it to a more general case

File: src/bashi/filter_software_dependency.py (hunk 3 above)
Comment (YuriiPerets) [reply src/bashi/filter_software_dependency.py]: Okay, I'm just keeping it for the draft version

**Comment by SimeonEhrig** [comment]
The PR is outdated. The several parts of this PR was implemented in the PRs #57, #58, #59 and #60.
